### PR TITLE
Markdown: add hashtags functionality

### PR DESCRIPTION
**Pull request by jiangyinzuo**
Some markdown editors or plugins, such as [Obsidian](https://help.obsidian.md/Editing+and+formatting/Tags) and [VSCode markdown-hashtags](https://github.com/vanadium23/markdown-hashtags), support parsing #hashtags and navigation. I have modified the markdown parser to support hashtag syntax parsing.

### CODE REVIEW

**Comment by jiangyinzuo** [comment]
Thanks for your tips! Now the polished PR is ready.

File: parsers/markdown.c [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include "parse.h"
 #include "read.h"
 #include "vstring.h"
+#include "utf8_str.h"
```
Comment (masatake) [inline parsers/markdown.c:31]: The changes for parsers/markdown.c should be a part of the second commit.

File: .gitignore [hunk 2]
```diff
@@ -21,6 +21,7 @@
 /Makefile
 /packcc
 /packcc.exe
+.cache
```
Comment (masatake) [inline .gitignore:24]: The change for .gitignore should be in a separate commit.

**Review by masatake — CHANGES_REQUESTED** [review]

File: parsers/markdown.c [hunk 3]
```diff
@@ -58,18 +59,25 @@ static kindDefinition MarkdownKinds[] = {
...
+	{ true, 'h', "hashtag",       "hashtag"},
 };
 
-static fieldDefinition MarkdownFields [] = {
+static fieldDefinition MarkdownFields[] = {
 	{
-	  .enabled     = false,
-	  .name        = "sectionMarker",
-	  .description = "character used for declaring section(#, ##, =, or -)",
-	},
```
Comment (masatake) [inline parsers/markdown.c:67]: It seems that the original code uses tab characters for indentation. Let's do the same in this pull request.

File: parsers/markdown.c [hunk 4]
```diff
@@ -259,6 +260,93 @@ static void notifyEndOfCodeBlock (markdownSubparser *m)
 	}
 }
 
+typedef enum {
+	HT_SPACE_FOUND,
```
Comment (masatake) [inline parsers/markdown.c:272]: Could you use HTAG_ as a prefix?
"HT_" makes me remember HT_ macros defined in main/htable.h.

File: main/utf8_str.c [hunk 5]
```diff
@@ -0,0 +1,80 @@
+/*
+*
+*   Copyright (c) 2023, Yinzuo Jiang
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   Defines functions for UTF-8 string manipulation.
+*/
+
```
Comment (masatake) [inline main/utf8_str.c:11]: Could you add the following comment?

```
/* utf8_strlen is derived from parsers/rst.c */

```

Comment (jiangyinzuo) [reply main/utf8_str.c:11]: Now added

File: .gitignore (hunk 2 above)
Comment (jiangyinzuo) [reply .gitignore:24]: Here I add a new separate commit `gitignore: add .cache and compile_commands.json`

File: parsers/markdown.c (hunk 1 above)
Comment (jiangyinzuo) [reply parsers/markdown.c:31]: fixed

File: parsers/markdown.c (hunk 3 above)
Comment (jiangyinzuo) [reply parsers/markdown.c:67]: Now changed in the same way.

File: parsers/markdown.c [hunk 6]
```diff
@@ -259,6 +268,94 @@ static void notifyEndOfCodeBlock (markdownSubparser *m)
 	}
 }
 
+typedef enum {
```
Comment (jiangyinzuo) [inline parsers/markdown.c:271]: Thank you for your advice. I renamed the prefix to `HTAG_`

File: parsers/markdown.c [hunk 7]
```diff
@@ -259,6 +268,94 @@ static void notifyEndOfCodeBlock (markdownSubparser *m)
...
+
+			int length = pos - start;
+			if (hasNonNumericalChar && length > 0)
+			{
+				vString *tag =
+					vStringNewNInit ((const char *) (&(line[start])), length);
+				int r = makeMarkdownTag (tag, K_HASHTAG, false);
+				attachParserFieldToCorkEntry (r,
+					MarkdownFields[F_HASHTAG].ftype, "#");
+				nestingLevelsPush (nestingLevels, r);
```
Comment (masatake) [inline parsers/markdown.c:336]: (review comment Group 2)

Pushed but never poped :-)


The `nestingLevels` is a variable for filling the scope field:

```
$ cat /tmp/input.md 
# section 1

## sub section 2

## sub section 3

# section A

## sub section B

## sub section C
$ ctags -o - --sort=no --fields-Markdown='{sectionMarker}' input.md
section 1	input.md	/^# section 1$/;"	c	sectionMarker:#
sub section 2	input.md	/^## sub section 2$/;"	s	chapter:section 1	sectionMarker:#
sub section 3	input.md	/^## sub section 3$/;"	s	chapter:section 1	sectionMarker:#
section A	input.md	/^# section A$/;"	c	sectionMarker:#
sub section B	input.md	/^## sub section B$/;"	s	chapter:section A	sectionMarker:#
sub section C	input.md	/^## sub section C$/;"	s	chapter:section A	sectionMarker:#
```

In my understanding, a hashtag doesn't make a scope.
So you don't have to push the tag (a value returned from `makeMarkdownTag`) for a hashtag to the `nestingLevels`. 

Your change makes one tag `a/b` for the input "#a/b".
If I make a pull request for the same purpose,
for aligning this pull request to the design of Universal Ctags, I will make two tags (`a` and `b`), and one full-qualified extra tag (`a/b`) for the input. `b` may have `hashtag:a` as the value for its scope field:
```
a	input.md	/^... #a/b ...$/;"	h
b	input.md	/^... #a/b ...$/;"	h	hashtag:a
```

This may help a client tool show a "tree view" of tags.

The full-qualified extra tag is emitted only when --extras=+q option is given.
With the option, in addition to the above `a` and `b`, ctags with my imaginary pull request, will emit:
```
a/b	input.md	/^... #a/b ...$/;"	h	extras:qualified
```

The `nestingLevels` is useful to tell ctags the relation between `a` and `b`; making `b` in the environment where `a` is pushed to the `nestingLevels`, ctags fills the scope field of `b` with `a` automatically. In addition, ctags also makes `a/b`.

Anyway, the current pull request is enough useful; yes, you are the user!
If you are interested in the fine-grained tagging, we can try implementing it after merging this:)

File: Units/parser-markdown.r/hashtags-utf8.d/expected.tags [hunk 8]
```diff
@@ -0,0 +1,23 @@
+标签	input.md	/^#标签 #태그 #وسم   $/;"	h	hashtag:#
+태그	input.md	/^#标签 #태그 #وسم   $/;"	h	hashtag:#
+وسم	input.md	/^#标签 #태그 #وسم   $/;"	h	end:1	hashtag:#
+t1	input.md	/^#t1 #t2#invalid#invalid2$/;"	h	end:1	hashtag:#
+t2	input.md	/^#t1 #t2#invalid#invalid2$/;"	h	end:6	hashtag:#
```
Comment (masatake) [inline Units/parser-markdown.r/hashtags-utf8.d/expected.tags:5]: (review comment Group 2)
`end:` shows an unexpected value because you calls `nestingLevelsPush`.

**Review by masatake — CHANGES_REQUESTED** [review]
Could you add the original comment of this pull request to the commit log? It helps a person understand this pull request.

```
Some markdown editors or plugins, such as [Obsidian](https://help.obsidian.md/Editing+and+formatting/Tags) and [VSCode markdown-hashtags](https://github.com/vanadium23/markdown-hashtags), support parsing #hashtags and navigation. I have modified the markdown parser to support hashtag syntax parsing.
```

**Comment by masatake** [comment]
The first commit and the third commit are ready to merge. Thank you.

**Comment by jiangyinzuo** [comment]
> The first commit and the third commit are ready to merge. Thank you. If you are comfortable, I will cherry-pick the first and third ones immediately. In that case, let me know.

It's OK

File: parsers/markdown.c (hunk 7 above)
Comment (jiangyinzuo) [reply parsers/markdown.c:336]: Thank you for your meticulousness and patience, I previously didn't understand the details of the code thoroughly enough. I will  attempt to implement it later on.

File: Units/parser-markdown.r/hashtags-utf8.d/input.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+#9_
+#-_-
+#t/s1 #t/s2
+#t4-5-6+excluded  #t7;excluded
+text#text #hashtag text#text
+#included
+    #excluded1
+	#excluded2
+   #included
```
Comment (masatake) [inline Units/parser-markdown.r/hashtags-utf8.d/input.md:15]: This comment is nothing to do with this pull request.
`#included` at line 15 is tagged as `c`hapter. This may be a bug of the original code.

Comment (masatake) [reply Units/parser-markdown.r/hashtags-utf8.d/input.md:15]: I opened #3748.

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
This one is ready for merging.
Could you give me a time before merging?

I would like to try solving https://github.com/universal-ctags/ctags/issues/3748 before merging this. If I find fixing #3748 may take longer time, I will merge this pull request first.

**Comment by jiangyinzuo** [comment]
> This one is ready for merging. Could you give me a time before merging?
> 
> I would like to try solving #3748 before merging this. If I find fixing #3748 may take longer time, I will merge this pull request first.

It's OK, and it is better to cherry-pick the first and third commits. I plan to open a new PR for **qualified hashtag** feature.

**Comment by jiangyinzuo** [comment]
I force-pushed again. Now the first commit is duplicated with #3749  and the third commit implements the features for `qualified hashtags` and `subtags`

**Comment by masatake** [comment]
Could you rebase this pull request on the latest master branch?
You must drop [dc0c700](https://github.com/universal-ctags/ctags/pull/3747/commits/dc0c700bcfab8ed1f827314b1539848a838ced9e) from this pull request.

**Comment by jiangyinzuo** [comment]
> Could you rebase this pull request on the latest master branch? You must drop [dc0c700](https://github.com/universal-ctags/ctags/pull/3747/commits/dc0c700bcfab8ed1f827314b1539848a838ced9e) from this pull request.

Done. Could you please review the second commit again?

File: parsers/markdown.c [hunk 10]
```diff
@@ -57,6 +60,8 @@ static kindDefinition MarkdownKinds[] = {
 	{ true, 'T', "l4subsection",  "level 4 sections" },
 	{ true, 'u', "l5subsection",  "level 5 sections" },
 	{ true, 'n', "footnote",      "footnotes" },
+	{ true, 'h', "hashtag",       "hashtags"},
+	{ true, 'H', "subtag",       "subtags in hashtags"},
```
Comment (masatake) [inline parsers/markdown.c:64]: Do you think we should have both hashtag and subtag?
Having only hashtag is not enough?

**Comment by masatake** [comment]
We are going the wrong way.

Could you push the original version to that the sub-hashtag was not introduced yet?
I will merge the original version.
We can work on the sub-hashtag after merging the original version.

**Comment by masatake** [comment]
BTW, I found you are using your `master` branch. I strongly recommend using a topic branch.
See https://git-scm.com/book/en/v2/Git-Branching-Branching-Workflows

So you can track the change on the master branch of the upstream and develop new features simultaneously.

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
Thank you!

**Comment by masatake** [comment]
I have some advises about tagging `#a/b`.

See php.c:
```
static scopeSeparator PhpGenericSeparators [] = {
        { K_NAMESPACE        , NAMESPACE_SEPARATOR },
        { KIND_WILDCARD_INDEX, "::" },
};
```

You can do the same as the PHP parser does:
```
static scopeSeparator MarkdownHashtagSeparators [] = {
        { K_HASHTAG        , "/" },  /* If a hashtag is the parent. */
        { KIND_WILDCARD_INDEX, "#" }, /* If something other is the parent. */
};
...
static kindDefinition MarkdownKinds[] = {
        { true, 'c', "chapter",       "chapters"},
        { true, 's', "section",       "sections" },
        { true, 'S', "subsection",    "level 2 sections" },
        { true, 't', "subsubsection", "level 3 sections" },
        { true, 'T', "l4subsection",  "level 4 sections" },
        { true, 'u', "l5subsection",  "level 5 sections" },
        { true, 'n', "footnote",      "footnotes" },
        { true, 'h', "hashtag",       "hashtags"
           ATTACH_SEPARATORS(MarkdownHashtagSeparators)}},

```

With this change, ctags generate full-qualified tags like `#a/b` with `--extras=+qufalifed` if you parser does:

- making a tag for `a` with the hashtag kind, and
- making a tag for `b` with the hashtag kind and with setting `a` to `b`'s scope.

When making a tag for `a`, makeTagEntry or something returns an integer.
We call it a corkIndex for `a`.
Set the corkIndex for `a` to `extensionFields.scopeIndex` of `b` when making a tag for `b`.
See makeMarkdownTag.
